lib/internal/wasi.js

```diff
@@ -0,0 +1,1301 @@
+            case rstats.isCharacterDevice():
+              filetype = WASI_FILETYPE_CHARACTER_DEVICE;
+              break;
+            case rstats.isDirectory():
+              filetype = WASI_FILETYPE_DIRECTORY;
+              break;
+            case rstats.isFIFO():
+              filetype = WASI_FILETYPE_SOCKET_STREAM;
+              break;
+            case rstats.isFile():
+              filetype = WASI_FILETYPE_REGULAR_FILE;
+              break;
+            case rstats.isSocket():
+              filetype = WASI_FILETYPE_SOCKET_STREAM;
+              break;
+            case rstats.isSymbolicLink():
+              filetype = WASI_FILETYPE_SYMBOLIC_LINK;
+              break;
+            default:
+              filetype = WASI_FILETYPE_UNKNOWN;
+              break;
+          }
+          this.view.setUint8(bufPtr, filetype);
+          bufPtr += 1;
+          bufPtr += 3; // padding
+          Buffer.from(this.memory.buffer)
+            .write(entry.name, bufPtr, bufLen - bufPtr);
+          bufPtr += Buffer.byteLength(entry.name);
+          bufPtr += (8 % bufPtr); // padding
+        }
+        const bufused = bufPtr - startPtr;
+        this.view.setUint32(bufusedPtr, bufused, true);
+        return WASI_ESUCCESS;
+      }),
+      fd_renumber: wrap((from, to) => {
+        CHECK_FD(from, 0);
+        CHECK_FD(to, 0);
+        fs.closeSync(this.FD_MAP.get(from).real);
+        this.FD_MAP.set(from, this.FD_MAP.get(to));
+        this.FD_MAP.delete(to);
+        return WASI_ESUCCESS;
+      }),
+      fd_seek: wrap((fd, offset, whence, newOffsetPtr) => {
+        const stats = CHECK_FD(fd, WASI_RIGHT_FD_SEEK);
+        const newOffset = binding.seek(stats.real, offset, {
+          [WASI_WHENCE_CUR]: SEEK_CUR,
+          [WASI_WHENCE_END]: SEEK_END,
+          [WASI_WHENCE_SET]: SEEK_SET,
+        }[whence]);
+        if (typeof newOffset === 'number') { // errno
+          throw newOffset;
+        }
+        this.refreshMemory();
+        this.view.setBigUint64(newOffsetPtr, newOffset, true);
+        return WASI_ESUCCESS;
+      }),
+      fd_tell: wrap((fd, offsetPtr) => {
+        const stats = CHECK_FD(fd, WASI_RIGHT_FD_TELL);
+        const currentOffset = binding.seek(stats.real, 0n, SEEK_CUR);
+        if (typeof currentOffset === 'number') { // errno
+          throw currentOffset;
+        }
+        this.refreshMemory();
+        this.view.setBigUint64(offsetPtr, currentOffset, true);
+        return WASI_ESUCCESS;
+      }),
+      fd_sync: wrap((fd) => {
+        const stats = CHECK_FD(fd, WASI_RIGHT_FD_SYNC);
+        fs.fsyncSync(stats.real);
+        return WASI_ESUCCESS;
+      }),
+      path_create_directory: wrap((fd, pathPtr, pathLen) => {
+        const stats = CHECK_FD(fd, WASI_RIGHT_PATH_CREATE_DIRECTORY);
+        if (!stats.path) {
+          return WASI_EINVAL;
+        }
+        this.refreshMemory();
+        const p = Buffer.from(this.memory.buffer, pathPtr, pathLen).toString();
```

Comment:
It’s really pretty unfortunate that we need to assume UTF-8 for all paths *and* we lose the advantage of `mkdirat()` of not having a race condition based on the path of `fd`… that might also be worth a `// TODO` comment?